src/server/protocol.ts

```diff
@@ -2159,6 +2159,7 @@ namespace ts.server.protocol {
     export interface CompletionEntryIdentifier {

```

Comment:
Is the only reason to keep `source` around backward compatibility at this point? From what I gather, it was a previous attempt at having a unique identifier for the item, but a simple string wasn't enough, hence `data`. 

Another question for backward compatibility - would it make sense, instead of creating a new property, to serialize `data` into some sort of string and send that data back in `source`? This way, clients may not have to make a code change to accommodate the new data? I can see that this would be super ugly though.

Reply:
> From what I gather, it was a previous attempt at having a unique identifier for the item, but a simple string wasn't enough, hence `data`.

This is correct.

> Is the only reason to keep `source` around backward compatibility at this point?

This is not yet true, because we also use `source` to disambiguate `this.`-inserting property completions from locals/globals with the same name as that property. But, it would be easy to move that info from `source` to `data` if we wanted to, so basically yes, back compat is the main reason I’m keeping it for now.

> Another question for backward compatibility - would it make sense, instead of creating a new property, to serialize `data` into some sort of string and send that data back in `source`?

This would have been an option had VS Code not started including `source` in their completion list UI (see https://github.com/microsoft/TypeScript/issues/42752#issuecomment-778519797). I believe if we were to change it to something JSON-like, it would disappear from the UI, which is of course better than rendering as raw JSON, but the current consensus is that showing an approximate path-like `source` for auto imports is better than showing nothing. So in short, VS Code has taken a dependency, for now, on the existing format of `source`.

Regarding backward compatibility in general, I asked in #42871 how long we’d need to allow clients to send us `source` and not `data`, which you might be able to comment on.

Reply:
I see, thanks!

I've been thinking about your last question since Friday and still don't have a satisfactory answer. Short answer: the de facto policy is "forever" - we don't have a formal versioning or compatibility story for the protocol. Many users decide to stay on older versions of VS for some reason or other, and many users always upgrade to the latest TS, and there's some overlap between the two, so we try to be as compatible as possible. If it becomes a big maintenance burden, it may be reasonable to introduce some breaking changes.

Note that I think your question not only applies to the protocol (which has implications for editors), but also to the API as well (which has implications for plugin developers). Which Daniel would have more insight on.

Reply:
Changes look good. Thanks for answering my questions!